### PR TITLE
Implement zip on primitive lists. Closes #308.

### LINKED ISSUE #308: Implement zip on primitive lists.
Signature should be as follows:

`ListIterable<<type><type>Pair> zip(<Name>Iterable iterable);`

### PR DESCRIPTION
**Pull request by donraab**
Signed-off-by: Donald Raab <Donald.Raab@bnymellon.com>

### CODE REVIEW

File: eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg [hunk 1]
```diff
@@ -397,6 +402,32 @@ final class Immutable<name>SingletonList implements Immutable<name>List, Seriali
...
+     * @since 9.1.
+     */
+    @Override
+    public ImmutableList\<<name><name>Pair> zip<name>(<name>List list)
+    {
+        if (list.isEmpty())
+        {
+            return Lists.immutable.empty();
+        }
+        return Lists.immutable.with(PrimitiveTuples.pair(this.get(0), list.get(0)));
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg]: this.get(0) should be replaced by this.element1

File: eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg [hunk 2]
```diff
@@ -397,6 +402,32 @@ final class Immutable<name>SingletonList implements Immutable<name>List, Seriali
...
+     * @since 9.1.
+     */
+    @Override
+    public \<T> ImmutableList\<<name>ObjectPair\<T\>> zip(ListIterable\<T> list)
+    {
+        if (list.isEmpty())
+        {
+            return Lists.immutable.empty();
+        }
+        return Lists.immutable.with(PrimitiveTuples.pair(this.get(0), list.get(0)));
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg]: this.get(0) should be replaced by this.element1

File: eclipse-collections-code-generator/src/main/resources/impl/list/mutable/unmodifiablePrimitiveList.stg [hunk 3]
```diff
@@ -237,6 +241,22 @@ public final class Unmodifiable<name>List
     {
         throw new UnsupportedOperationException("subList not yet implemented!");
     }
+
+    /**
+     * @since 9.1.
+     */
+    public MutableList\<<name><name>Pair> zip<name>(<name>List list)
+    {
+        return this.getMutable<name>List().zip<name>(list);
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/list/mutable/unmodifiablePrimitiveList.stg]: Should we return an UnmodifiableList?

File: eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg [hunk 4]
```diff
@@ -288,6 +293,64 @@ public abstract class AbstractImmutable<name>ListTestCase extends AbstractImmuta
...
+                zipLess);
+        ImmutableList\<<name><name>Pair> zipMore = list2.zip<name>(list1);
+        Assert.assertEquals(
+                Lists.immutable.with(
+                    PrimitiveTuples.pair(<["1", "1"]:(literal.(type))(); separator=", ">)),
+                zipMore);
+        ImmutableList\<<name><name>Pair> zipEmpty1 = list1.zip<name>(this.newWith());
+        Assert.assertTrue(zipEmpty1.isEmpty());
+        ImmutableList\<<name><name>Pair> zipEmpty2 = this.newWith().zip<name>(list1);
+        Assert.assertTrue(zipEmpty2.isEmpty());
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg]: There needs to be a test case for list1.size() = list2.size()

File: eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg [hunk 5]
```diff
@@ -288,6 +293,64 @@ public abstract class AbstractImmutable<name>ListTestCase extends AbstractImmuta
...
+                zipLess);
+        ImmutableList\<<name>ObjectPair\<String\>> zipMore = list2.zip(list3);
+        Assert.assertEquals(
+                Lists.immutable.with(
+                    PrimitiveTuples.pair(<(literal.(type))("1")>, "1")),
+                zipMore);
+        ImmutableList\<<name>ObjectPair\<String\>> zipEmpty1 = list1.zip(Lists.immutable.empty());
+        Assert.assertTrue(zipEmpty1.isEmpty());
+        ImmutableList\<<name>ObjectPair\<String\>> zipEmpty2 = this.newWith().zip(Lists.immutable.with("1", "2"));
+        Assert.assertTrue(zipEmpty2.isEmpty());
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg]: Tests for equal size

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CharAdapter.java [hunk 6]
```diff
@@ -531,6 +537,38 @@ public int hashCode()
...
+            target.add(PrimitiveTuples.pair(this.get(i), list.get(i)));
+        }
+        return target.toImmutable();
+    }
+
+    /**
+     * @since 9.1.
+     */
+    @Override
+    public <T> ImmutableList<CharObjectPair<T>> zip(ListIterable<T> list)
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CharAdapter.java]: Should the scope of T be expanded `<T extends ?>` ?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CodePointAdapter.java [hunk 7]
```diff
@@ -705,6 +711,38 @@ public int hashCode()
...
+            target.add(PrimitiveTuples.pair(this.get(i), list.get(i)));
+        }
+        return target.toImmutable();
+    }
+
+    /**
+     * @since 9.1.
+     */
+    @Override
+    public <T> ImmutableList<IntObjectPair<T>> zip(ListIterable<T> list)
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CodePointAdapter.java]: Should scope of T be expanded `<T extends ?>`

File: eclipse-collections-code-generator/src/main/resources/impl/list/mutable/unmodifiablePrimitiveList.stg (hunk 3 above)
Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/impl/list/mutable/unmodifiablePrimitiveList.stg]: I believe this is consistent with other methods.  For example, select:

```
    @Override
    public Mutable<name>List select(<name>Predicate predicate)
    {
        return this.getMutable<name>List().select(predicate);
    }
```

File: eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg (hunk 4 above)
Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg]: Zip with the same list should cover this case.

File: eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg (hunk 5 above)
Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg]: Zip with the same list should cover this case.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CharAdapter.java (hunk 6 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CharAdapter.java]: Not sure what you are suggesting in this case.  This is consistent with definition in ListIterable.

```    
    @Override
    <S> ListIterable<Pair<T, S>> zip(Iterable<S> that);
```

File: eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg (hunk 1 above)
Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg]: Makes sense.  Fixed.

File: eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg (hunk 2 above)
Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveSingletonList.stg]: Makes sense.  Fixed.

File: eclipse-collections-code-generator/src/main/resources/impl/list/mutable/unmodifiablePrimitiveList.stg (hunk 3 above)
Comment (nikhilnanivadekar) [reply eclipse-collections-code-generator/src/main/resources/impl/list/mutable/unmodifiablePrimitiveList.stg]: ok makes sense.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CharAdapter.java (hunk 6 above)
Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/string/immutable/CharAdapter.java]: ok makes sense.

File: eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg (hunk 4 above)
Comment (nikhilnanivadekar) [reply eclipse-collections-code-generator/src/main/resources/test/list/immutable/abstractImmutablePrimitiveListTestCase.stg]: Ah! Sorry I missed that in this file.

File: eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg [hunk 8]
```diff
@@ -699,6 +705,38 @@ final class Immutable<name>ArrayList
...
+        return target.toImmutable();
+    }
+
+    /**
+     * @since 9.1.
+     */
+    @Override
+    public \<T> ImmutableList\<<name>ObjectPair\<T\>> zip(ListIterable\<T> list)
+    {
+        MutableList\<<name>ObjectPair\<T\>> target = Lists.mutable.empty();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg]: We should have a shortcut which checks for list.isEmpty() and returns. I am fine merging this PR if you want to do it as a part of next commit/PR. Please raise an issue and add "good for first issue" label.

Comment (motlin) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg]: I don't think it's necessary to special case isEmpty(). However, calling list.get() without checking RandomAccess makes this O(n^2) worst case.

Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg]: Good point.

**Pull request merged**